### ISSUE TITLE
Add support for nested splits

### ISSUE THREAD

**Issue by nad** [mortberg/cubicaltt#5]
Nested splits don't seem to be supported at the moment, and in the absence of mutual recursion the workarounds can be quite hairy:

```
module nestedSplits where

import bool
import nat

equalZero : nat -> bool =
  split
    zero  -> true
    suc n -> false

equalSuc
  (equal : nat -> nat -> bool) (m : nat) : nat -> bool =
  split
    zero  -> false
    suc n -> equal m n

equal : nat -> nat -> bool =
  split
    zero  -> equalZero
    suc n -> equalSuc equal n

lessComplicatedButRejected : nat -> nat -> bool =
  split
    zero -> split
      zero  -> true
      suc n -> false
    suc m -> split
      zero  -> false
      suc n -> lessComplicatedButRejected m n
```

**Comment by abooij** [comment]
This also makes it tedious to do recursion on e.g. the second argument of a function (which is quite useful for more complicated types).

**Comment by mortberg** [comment]
tulcod I'm not sure I understand what you mean. For example addition on nat is defined by recursion on the second argument: 

add (m : nat) : nat -> nat = split
  zero  -> m
  suc n -> suc (add m n)

**Comment by abooij** [comment]
@mortberg: right, but how would you define it by recursion on the first argument (aside from the fact that, in this case, that would be the same)?

edit: ah, I understand why my phrasing was not clear. i hope the problem is clear now.

**Comment by mortberg** [comment]
I would do it like this: 

```
add : nat -> nat -> nat = split
  zero -> \(n : nat) -> n
  suc m -> \(n : nat) -> suc (add m n)
```

So I still don't see what you mean @tulcod ...

**Comment by abooij** [comment]
hah, it seems I was unaware of the subtleties of the type syntax in `add (m : nat) : nat -> nat`. thanks.

**Comment by mortberg** [comment]
Anyway I think it would be nice to have split as an expression, however for doing HITs we need to know the type so they

**Comment by mortberg** [comment]
I added support for this now. For HITs we need to know the type of the split which means that the user has to provide it when the split is not at the toplevel. The syntax I came up with looks like this: 

```
equalNat : nat -> nat -> bool = split
    zero -> split@(nat -> bool) with
      zero  -> true
      suc n -> false
    suc m -> split@(nat -> bool) with
      zero  -> false
      suc n -> equalNat m n
```

Implementing this was very straightforward and I only had to change the parser and resolver. One can imagine more clever solutions where the user don't have to give the type. For example the resolver could try to keep track of the type while traversing the term or we could have different syntax for split for datatypes and HITs. Or maybe the ideal solution would be to have type inference and the type could maybe be inferred. However all of these solutions seem like too much work just to avoid writing the type. At least with this new syntax one doesn't have to make a new declaration every time one wants to do a nested split.
